src/NHibernate/Impl/SessionFactoryImpl.cs

```diff
@@ -214,6 +214,22 @@ public SessionFactoryImpl(Configuration cfg, IMapping mapping, Settings settings
 				log.Warn(ex, "Dialect does not provide DataBaseSchema, but keywords import or auto quoting is enabled.");
 			}
 
+			#region Serialization info
+
+			name = settings.SessionFactoryName;
+			try
+			{
+				uuid = (string)UuidGenerator.Generate(null, null);
+			}
+			catch (Exception ex)
+			{
+				throw new AssertionFailure("Could not generate UUID", ex);
+			}
+
+			SessionFactoryObjectFactory.AddInstance(uuid, name, this, properties);
+
+			#endregion
```

Comment:
This fix, moving name/uuid registration at the beginning of the session factory construction, seems to work and does not cause any old test failure, but is quite debatable.

It allows deserialization to resolve a session factory not yet fully built. And this occurs in this Prevalence cache case, meaning it will recompute its cached entities keys hashcodes with a session factory not yet fully built.

This case is quite wicked. The persisters creation creates the caches, so we have to start their provider before creating the persisters.